File: packages/ruby-sdk/lib/strait/operations/webhooks.rb
# frozen_string_literal: true

module Strait
  module Operations
    # Service for webhook subscription and delivery management.
    class WebhooksService < BaseService
      # List webhook subscriptions.
      # GET /v1/webhooks/subscriptions
      def list_subscriptions(query: nil)
        _request(:get, "/v1/webhooks/subscriptions", query: query)
      end

      # Create a webhook subscription.
      # POST /v1/webhooks/subscriptions
      def create_subscription(body)
        _request(:post, "/v1/webhooks/subscriptions", body: body)
      end

      # Delete a webhook subscription.
      # DELETE /v1/webhooks/subscriptions/{id}
      def delete_subscription(id)
        _request(:delete, "/v1/webhooks/subscriptions/{id}", path_params: { "id" => id })
        nil
      end

      # List webhook deliveries.
      # GET /v1/webhooks/deliveries
      def list_deliveries(query: nil)
        _request(:get, "/v1/webhooks/deliveries", query: query)
      end

      # Get a webhook delivery by ID.
      # GET /v1/webhooks/deliveries/{id}
      def get_delivery(id)
        _request(:get, "/v1/webhooks/deliveries/{id}", path_params: { "id" => id })
      end

      # Retry a webhook delivery.
      # POST /v1/webhooks/deliveries/{id}/retry
      def retry_delivery(id)
        _request(:post, "/v1/webhooks/deliveries/{id}/retry", path_params: { "id" => id })
      end
    end
  end
end

File: packages/python-sdk/strait/operations/webhooks.py
"""Webhook management operations."""

from __future__ import annotations

from typing import Any

from strait.operations._base import AsyncBaseService, BaseService


class WebhooksService(BaseService):
    def list_subscriptions(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return self._request("GET", "/v1/webhooks/subscriptions", query=query)

    def create_subscription(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/webhooks/subscriptions", body=body)

    def delete_subscription(self, id: str) -> dict[str, Any]:
        return self._request(
            "DELETE", "/v1/webhooks/subscriptions/{id}", path_params={"id": id},
        )

    def list_deliveries(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return self._request("GET", "/v1/webhooks/deliveries", query=query)

    def get_delivery(self, id: str) -> dict[str, Any]:
        return self._request(
            "GET", "/v1/webhooks/deliveries/{id}", path_params={"id": id},
        )

    def retry_delivery(self, id: str) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/webhooks/deliveries/{id}/retry", path_params={"id": id},
        )


class AsyncWebhooksService(AsyncBaseService):
    async def list_subscriptions(
        self, *, query: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        return await self._request("GET", "/v1/webhooks/subscriptions", query=query)

    async def create_subscription(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/webhooks/subscriptions", body=body)

    async def delete_subscription(self, id: str) -> dict[str, Any]:
        return await self._request(
            "DELETE", "/v1/webhooks/subscriptions/{id}", path_params={"id": id},
        )

    async def list_deliveries(
        self, *, query: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        return await self._request("GET", "/v1/webhooks/deliveries", query=query)

    async def get_delivery(self, id: str) -> dict[str, Any]:
        return await self._request(
            "GET", "/v1/webhooks/deliveries/{id}", path_params={"id": id},
        )

    async def retry_delivery(self, id: str) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/webhooks/deliveries/{id}/retry", path_params={"id": id},
        )
